File: Data Structures & Algorithms/n-queens/submission-0.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        char[][] b = new char[n][n];
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                b[i][j] = '.';
            }
        }

        bt(b,res,0);
        return res;
    }

    public void bt(char[][] b, List<List<String>> res, int r){
        if(r == b.length){
            List<String> cp = new ArrayList<>();
            for(char[] f:b){
                cp.add(new String(f));
            }
            res.add(cp);
            return;
        }

        for(int c = 0;c<b.length;c++){
            if(isafe(b,r,c)){
                b[r][c] = 'Q';
                bt(b,res,r+1);
                b[r][c] = '.';
            }
        }
    }

    public boolean isafe(char[][] b, int r, int c){
        for(int i=r-1;i>=0;i--){
            if(b[i][c] == 'Q') return false;
        }
        for(int i = r-1, j = c-1; i>=0 && j>=0; i--,j--){
            if(b[i][j] == 'Q') return false;
        }
        for(int i = r-1, j = c+1; i>=0 && j<b.length;i--,j++){
            if(b[i][j] == 'Q') return false;
        }
        return true;
    }
}

File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-0.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int l = *max_element(weights.begin(),weights.end());
        int r = accumulate(weights.begin(),weights.end(),0);
        int res = r;

        while(l <= r){
            int mid = (l+r)/2;
            if(cap(weights,days,mid)){
                res = min(res,mid);
                r = mid-1;
            }else{
                l = mid+1;
            }
        }
        return res;
    }

private:
    bool cap(vector<int>& w, int d, int c){
        int s = 1;
        int cc = c;

        for(int i:w){
            if(cc-i < 0){
                s++;
                if(s > d){
                    return false;
                }
                cc = c;
            }
            cc -= i;
        }
        return true;
    }
};